Ensure that LayerRenderJob objects are never copied

nyalldawson opened the pull request:
When a LayerRenderJob object is copied, it would create a new
copy-constructed QgsRenderContext member as a result. But we
store references to the LayerRenderJob's context in the
QgsMapLayerRenderer instances, so this potentially results
in using references to deleted objects.

In Qt5 builds this wasn't an issue, as LayerRenderJob objects
weren't being copied. But Qt6's revamped containers WERE
detaching and copying LayerRenderJob objects, resulting
in crashes in all the map renderer tests.

So... explicitly block copying of LayerRenderJob to fix this.
And then use std::vector instead of QList for storage of job
lists as we can't use Qt containers with non-copy-constructible
classes.

(This is a step forward anyway, because the ownership of the
raw pointers in LayerRenderJob has always been somewhat opaque,
and by avoiding copies of LayerRenderJob we have the opportunity
to later clean this up and make them unique_ptrs were appropriate)

Fixes all map renders crashing on Qt6 builds

nyalldawson committed:
Ensure that LayerRenderJob objects are never copied

When a LayerRenderJob object is copied, it would create a new
copy-constructed QgsRenderContext member as a result. But we
store references to the LayerRenderJob's context in the
QgsMapLayerRenderer instances, so this potentially results
in using references to deleted objects.

In Qt5 builds this wasn't an issue, as LayerRenderJob objects
weren't being copied. But Qt6's revamped containers WERE
detaching and copying LayerRenderJob objects, resulting
in crashes in all the map renderer tests.

So... explicitly block copying of LayerRenderJob to fix this.
And then use std::vector instead of QList for storage of job
lists as we can't use Qt containers with non-copy-constructible
classes.

(This is a step forward anyway, because the ownership of the
raw pointers in LayerRenderJob has always been somewhat opaque,
and by avoiding copies of LayerRenderJob we have the opportunity
to later clean this up and make them unique_ptrs were appropriate)

Fixes all map renders crashing on Qt6 builds

rouault commented on src/core/maprenderer/qgsmaprendererjob.cpp:
```diff
@@ -644,19 +700,24 @@ LayerRenderJobs QgsMapRendererJob::prepareSecondPassJobs( LayerRenderJobs &first
+    secondPassJobs.emplace_back( LayerRenderJob() );
+    LayerRenderJob &job2 = secondPassJobs.back();
+
+#if 0
```
isn't some initialization of job2 lacking now that job2 = job is disabled ?

nyalldawson replied:
Oops, had disabled this temporarily..

rouault commented on src/core/maprenderer/qgsmaprendererjob.cpp:
```diff
@@ -720,15 +781,15 @@ LabelRenderJob QgsMapRendererJob::prepareLabelingJob( QPainter *painter, QgsLabe
+void QgsMapRendererJob::cleanupJobs( std::vector<LayerRenderJob> &jobs )
 {
-  for ( LayerRenderJobs::iterator it = jobs.begin(); it != jobs.end(); ++it )
+  for ( auto it = jobs.begin(); it != jobs.end(); ++it )
```
``for ( auto &job : jobs )`` would probably do it

rouault commented on src/core/maprenderer/qgsmaprendererjob.cpp:
```diff
@@ -836,7 +896,7 @@ QImage QgsMapRendererJob::composeImage(
   int i = 0;
 #endif
-  for ( LayerRenderJobs::const_iterator it = jobs.constBegin(); it != jobs.constEnd(); ++it )
+  for ( auto it = jobs.begin(); it != jobs.end(); ++it )
```
``for ( auto &job : jobs )`` would probably do it

rouault commented on src/core/maprenderer/qgsmaprendererjob.cpp:
```diff
@@ -879,7 +939,7 @@ QImage QgsMapRendererJob::composeImage(
 
   // render any layers with the renderAboveLabels flag now
-  for ( LayerRenderJobs::const_iterator it = jobs.constBegin(); it != jobs.constEnd(); ++it )
+  for ( auto it = jobs.begin(); it != jobs.end(); ++it )
```
``for ( auto &job : jobs )`` would probably do it

rouault commented on src/core/maprenderer/qgsmaprendererparalleljob.cpp:
```diff
@@ -125,9 +125,9 @@ void QgsMapRendererParallelJob::cancelWithoutBlocking()
 
   mLabelJob.context.setRenderingStopped( true );
-  for ( LayerRenderJobs::iterator it = mLayerJobs.begin(); it != mLayerJobs.end(); ++it )
+  for ( auto it = mLayerJobs.begin(); it != mLayerJobs.end(); ++it )
```
``for ( auto &job : mLayerJobs )`` would probably do it

rouault commented on src/core/maprenderer/qgsmaprendererparalleljob.cpp:
```diff
@@ -224,8 +224,7 @@ void QgsMapRendererParallelJob::renderLayersFinished()
 
-  LayerRenderJobs::const_iterator it = mLayerJobs.constBegin();
-  for ( ; it != mLayerJobs.constEnd(); ++it )
+  for ( auto it = mLayerJobs.begin(); it != mLayerJobs.end(); ++it )
```
``for ( auto &job : mLayerJobs )`` would probably do it

nyalldawson committed:
Fix partial copy of LayerRenderJob for second pass job

nyalldawson committed:
Cleanup loops

rouault commented on src/core/maprenderer/qgsmaprendererjob.cpp:
```diff
@@ -703,13 +703,26 @@ std::vector< LayerRenderJob > QgsMapRendererJob::prepareSecondPassJobs( std::vec
+    job2.errors = job.errors;
+    job2.layerId = job.layerId;
+    job2.maskImage = job.maskImage;
+    job.maskJobs = job.maskJobs;
```
```suggestion
    job2.maskJobs = job.maskJobs;
```

nyalldawson committed:
Don't copy first pass LayerRenderJobs for second pass jobs

We only need to copy the render context and then associate the
second pass job with the original layer id. We don't use any
other values from the first pass job.

nyalldawson commented on src/core/maprenderer/qgsmaprendererjob.cpp:
```diff
@@ -703,13 +703,26 @@ std::vector< LayerRenderJob > QgsMapRendererJob::prepareSecondPassJobs( std::vec
+    // copy safe(?) things from first pass job
+    // TODO -- this list should be refined, there's a lot of pointer copies here
+    // which are likely unsafe
+    job2.setContext( std::make_unique< QgsRenderContext >( *job.context() ) );
```
I've cleaned this up now -- turns out it's only the context which needs setting and then the second pass job needs the layer association too. Everything else was unused in the second pass job.
